spec/unit/event-timeline.spec.js

```diff
@@ -0,0 +1,331 @@
+"use strict";
+var sdk = require("../..");
+var EventTimeline = sdk.EventTimeline;
+var utils = require("../test-utils");
+
+describe("EventTimeline", function() {
+    var roomId = "!foo:bar";
+    var userA = "@alice:bar";
+    var userB = "@bertha:bar";
+    var timeline;
+
+    beforeEach(function() {
+        utils.beforeEach(this);
+        timeline = new EventTimeline(roomId);
+
+        // mock RoomStates
+        timeline._startState = utils.mock(sdk.RoomState, "startState");
+        timeline._endState = utils.mock(sdk.RoomState, "endState");
+    });
+
+    describe("construction", function() {
+        it("getRoomId should get room id", function() {
+            var v = timeline.getRoomId();
+            expect(v).toEqual(roomId);
+        });
+    });
+
+    describe("initialiseState", function() {
+        it("should copy state events to start and end state", function() {
+            var events = [
+                utils.mkMembership({
+                    room: roomId, mship: "invite", user: userB, skey: userA,
+                    event: true,
+                }),
+                utils.mkEvent({
+                    type: "m.room.name", room: roomId, user: userB,
+                    event: true,
+                    content: { name: "New room" },
+                })
+            ];
+            timeline.initialiseState(events);
+            expect(timeline._startState.setStateEvents).toHaveBeenCalledWith(
+                events
+            );
+            expect(timeline._endState.setStateEvents).toHaveBeenCalledWith(
+                events
+            );
+        });
+
+        it("should raise an exception if called after events are added", function() {
+            var event =
+                utils.mkMessage({
+                    room: roomId, user: userA, msg: "Adam stole the plushies",
+                    event: true,
+                });
+
+            var state =
+                utils.mkMembership({
+                    room: roomId, mship: "invite", user: userB, skey: userA,
+                    event: true,
+                });
+
+            expect(function() { timeline.initialiseState(state); }).not.toThrow();
+            timeline.addEvent(event, false);
+            expect(function() { timeline.initialiseState(state); }).toThrow();
+        });
+    });
+
+    describe("paginationTokens", function() {
+        it("pagination tokens should start undefined", function() {
+            expect(timeline.getPaginationToken(true)).toBe(undefined);
+            expect(timeline.getPaginationToken(false)).toBe(undefined);
```

Comment:
I'd prefer if they started as `null` unless there is a good reason for them not to be. We typically initialise "no value" vars to `null` in the rest of the code, so if you see `undefined` errors it's most likely a typo.